src/resources/lib/modules/services.py

```diff
@@ -605,4 +626,34 @@ def wizard_set_samba(self):
         except Exception, e:
             self.oe.dbg_log('services::wizard_set_samba', 'ERROR: (%s)' % repr(e))
 
-
+    def do_sshpasswd(self, **kwargs):
+        try:
+            self.oe.dbg_log('system::do_sshpasswd', 'enter_function', 0)
+            if 'listItem' in kwargs:
+                self.set_value(kwargs['listItem'])
+            state = 1
+            options = {}
+            if self.struct['ssh']['settings']['ssh_autostart']['value'] == '1':
+                options['SSH_PASSWORD'] = '"%s"' % self.struct['ssh']['settings']['ssh_passwd']['value']
+            else:
+                state = 0
+            xbmcDialog = xbmcgui.Dialog()
+            ssh = subprocess.Popen(["passwd"], shell=False, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

Comment:
Wouldn't it be better to open the subprocess _after_ we have obtained the users input?

And if the user cancels the virtual keyboard (by clicking cancel) we proceed to set a blank password.

So I would suggest something like:

```
            xbmcDialog = xbmcgui.Dialog()
            newpwd = xbmcDialog.input(_(746))

            if not newpd:
#                xbmcDialog.ok(_(3222X), _(3222Y))
                self.oe.dbg_log('system::do_sshpasswd', 'exit_function - password unchanged', 0)
                return

            ssh = subprocess.Popen(["passwd"], shell=False, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

            readout1 = ssh.stdout.readline()
            ssh.stdin.write(newpwd + '\n')
            ssh.stdin.flush()

            readout2 = ssh.stdout.readline()
            ssh.stdin.write(newpwd + '\n')

            readout3 = ssh.stdout.readline()

            if "Bad password" in readout3:
                xbmcDialog.ok(_(32220), _(32221))
                self.oe.dbg_log('system::do_sshpasswd', 'exit_function password too weak', 0)
                return
            elif "Retype password" in readout3:
                xbmcDialog.ok(_(32222), _(32223))
...
```
etc.

I've commented out a dialog that could be shown when cancelling the password change, where `3222X` corresponds to `Cancelled`, and `3222Y` the message, something like `root password is unchanged.`, as it might be useful to confirm that the password is in fact not being changed... Alternatively, just cancel the keyboard and show nothing.

Reply:
agreed - don't think we need the dialog after a user cancel - seems like user will know they hit cancel and end result is no change, so no problem